File: windows/src/Slipstream.Core/Pairing/PairingSession.cs
using Slipstream.Core.Control;
using Slipstream.Core.Identity;

namespace Slipstream.Core.Pairing;

public enum PairingState
{
    AwaitingOffer,
    AwaitingConfirmation,
    Paired,
    Cancelled,
}

/// <summary>
/// One pairing attempt, on one device. Spec §4.
///
/// The six-digit code is derived from both certificate fingerprints and is never
/// transmitted — each side computes it and the user compares them by eye. Pairing
/// completes only on mutual confirmation: this device's user AND the remote.
/// </summary>
public sealed class PairingSession(DeviceIdentity localIdentity)
{
    private readonly Lock _gate = new();

    private PairOfferPayload? _offer;
    private string? _verifiedFingerprint;
    private bool _confirmedLocally;
    private bool _confirmedRemotely;

    public PairingState State { get; private set; } = PairingState.AwaitingOffer;

    /// <summary>The six digits to show the user. Null until an offer arrives.</summary>
    public string? Code { get; private set; }

    public PairedPeer? Result { get; private set; }

    /// <param name="verifiedFingerprint">
    /// Taken from the TLS certificate, never from <paramref name="offer"/>. The offer is
    /// peer-supplied text; the certificate is the only thing the handshake proves.
    /// </param>
    public void ReceiveOffer(PairOfferPayload offer, string verifiedFingerprint)
    {
        lock (_gate)
        {
            if (State != PairingState.AwaitingOffer) return;

            // A peer claiming a fingerprint it does not hold is either broken or hostile.
            if (!string.Equals(offer.Fingerprint, verifiedFingerprint, StringComparison.OrdinalIgnoreCase))
            {
                State = PairingState.Cancelled;
                return;
            }

            // Never pair with ourselves.
            if (string.Equals(verifiedFingerprint, localIdentity.Fingerprint, StringComparison.OrdinalIgnoreCase))
            {
                State = PairingState.Cancelled;
                return;
            }

            _offer = offer;
            _verifiedFingerprint = verifiedFingerprint;
            Code = PairingCode.Derive(localIdentity.Fingerprint, verifiedFingerprint);
            State = PairingState.AwaitingConfirmation;
        }
    }

    /// <summary>This device's user confirmed the codes match.</summary>
    public void ConfirmLocally()
    {
        lock (_gate)
        {
            if (State != PairingState.AwaitingConfirmation) return;

            _confirmedLocally = true;
            CompleteIfMutual();
        }
    }

    public void ReceiveRemoteConfirm()
    {
        lock (_gate)
        {
            if (State != PairingState.AwaitingConfirmation) return;

            _confirmedRemotely = true;
            CompleteIfMutual();
        }
    }

    public void Cancel()
    {
        lock (_gate)
        {
            if (State == PairingState.Paired) return;
            State = PairingState.Cancelled;
        }
    }

    private void CompleteIfMutual()
    {
        // A single-sided confirmation never pairs. Both users looked at the code.
        if (!_confirmedLocally || !_confirmedRemotely) return;

        Result = new PairedPeer(
            _offer!.DeviceId,
            _verifiedFingerprint!,
            _offer.Name,
            DateTimeOffset.UtcNow);

        State = PairingState.Paired;
    }
}

File: android/core/src/main/kotlin/com/slipstream/core/pairing/PairingSession.kt
package com.slipstream.core.pairing

import com.slipstream.core.identity.DeviceIdentity
import com.slipstream.core.identity.PairingCode
import kotlinx.serialization.Serializable

/** pairing.md §3: the state both devices' [PairingSession] step through, in lockstep. */
enum class PairingState {
    Idle,
    AwaitingConfirmation,
    Confirmed,
    Cancelled,
}

/** Wire shape of a `pair.offer` payload (pairing.md §3 example JSON). */
@Serializable
data class PairingOffer(
    val version: Int = 1,
    val deviceId: String,
    val name: String,
    val fingerprint: String,
)

/** What a successfully-confirmed pairing session produced: the remote device's identity as
 * it was verified during the exchange. The caller combines this with the X.509 certificate
 * obtained from the TLS handshake itself (never from the wire) to build a [com.slipstream.core.identity.PairedPeer]. */
data class PairingOutcome(
    val deviceId: String,
    val name: String,
    val fingerprint: String,
)

/**
 * pairing.md §3-§5: one side's half of the pairing state machine. Both devices run an
 * identical [PairingSession]; [PairingCoordinator] drives it over the wire.
 *
 * The critical invariant lives in [receiveOffer]: the code is derived from
 * [verifiedFingerprint] — the certificate fingerprint proven by the TLS handshake — never
 * from [PairingOffer.fingerprint], which is peer-supplied text and proves nothing by
 * itself. An offer whose claimed fingerprint disagrees with the verified one is rejected
 * (state -> [PairingState.Cancelled]) before a code is ever computed.
 *
 * Persistence (by the caller, via [PairingOutcome]) only ever happens once both
 * [confirmLocally] and [receiveConfirm] have been observed — a single-sided confirmation
 * leaves the session in [PairingState.AwaitingConfirmation] forever.
 */
class PairingSession(private val local: DeviceIdentity) {

    var state: PairingState = PairingState.Idle
        private set

    var code: String? = null
        private set

    var result: PairingOutcome? = null
        private set

    private var remoteOffer: PairingOffer? = null
    private var remoteFingerprint: String? = null
    private var localConfirmed = false
    private var remoteConfirmed = false

    /** This device's own offer, ready to send as the `pair.offer` payload. */
    fun localOffer(): PairingOffer = PairingOffer(
        deviceId = local.deviceId,
        name = local.displayName,
        fingerprint = local.fingerprint,
    )

    /**
     * Handles an incoming `pair.offer`. [verifiedFingerprint] must be the fingerprint of the
     * certificate actually presented during the TLS handshake — not anything read off the
     * wire. Rejects (cancels) if [offer]'s claimed fingerprint doesn't match it, before
     * deriving any code.
     */
    fun receiveOffer(offer: PairingOffer, verifiedFingerprint: String) {
        if (state == PairingState.Cancelled || state == PairingState.Confirmed) return

        if (offer.fingerprint != verifiedFingerprint) {
            state = PairingState.Cancelled
            code = null
            return
        }

        remoteOffer = offer
        remoteFingerprint = verifiedFingerprint
        code = PairingCode.derive(local.fingerprint, verifiedFingerprint)
        state = PairingState.AwaitingConfirmation
    }

    /** This device's user accepted the displayed code. */
    fun confirmLocally() {
        if (state != PairingState.AwaitingConfirmation) return
        localConfirmed = true
        tryFinish()
    }

    /** The remote device sent `pair.confirm`. */
    fun receiveConfirm() {
        if (state != PairingState.AwaitingConfirmation) return
        remoteConfirmed = true
        tryFinish()
    }

    /** Either side declined, or the remote sent `pair.cancel`, or the connection dropped. */
    fun cancel() {
        if (state == PairingState.Confirmed) return
        state = PairingState.Cancelled
        code = null
        result = null
    }

    private fun tryFinish() {
        // Persistence only ever happens with BOTH confirmations observed - a single-sided
        // confirm must never pair.
        if (!localConfirmed || !remoteConfirmed) return
        val offer = remoteOffer ?: return
        val fingerprint = remoteFingerprint ?: return
        result = PairingOutcome(deviceId = offer.deviceId, name = offer.name, fingerprint = fingerprint)
        state = PairingState.Confirmed
    }
}
